### DISCUSSION TITLE
Sudden error with mariaDB (no implicit conversion of Integer into String)

### DISCUSSION THREAD

**Discussion by wuarmin** [jeremyevans/sequel#1821] [General]
Hello,

I had a working connection to a MariaDB-instance, but suddenly the following error occurs, if I access a relation:

```ruby
db = Sequel.connect("mysql2://user:passwd@host:3306/mydb", :loggers => [Logger.new($stdout)])
db[:my_table] # here error occurs
```
```txt
TypeError: no implicit conversion of Integer into String
  /home/me/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sequel-5.52.0/lib/sequel/adapters/shared/mysql.rb:130:in `match'
  /home/me/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sequel-5.52.0/lib/sequel/adapters/shared/mysql.rb:130:in `server_version'
  /home/me/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sequel-5.52.0/lib/sequel/adapters/mysql2.rb:81:in `server_version'
  /home/me/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sequel-5.52.0/lib/sequel/adapters/shared/mysql.rb:796:in `supports_cte?'
  /home/me/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sequel-5.52.0/lib/sequel/dataset/sql.rb:1545:in `select_with_sql'
  /home/me/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sequel-5.52.0/lib/sequel/dataset/sql.rb:239:in `select_sql'
  /home/me/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sequel-5.52.0/lib/sequel/dataset/actions.rb:152:in `each'
  /home/me/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sequel-5.52.0/lib/sequel/dataset/actions.rb:51:in `block in all'
  /home/me/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sequel-5.52.0/lib/sequel/dataset/actions.rb:1010:in `_all'
  /home/me/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sequel-5.52.0/lib/sequel/dataset/actions.rb:51:in `all'
  /home/me/code/tou_etl_jobs/system/boot/telespirit.rb:11:in `block (2 levels) in <top (required)>'
```
I don't know what the problem is. The code was still working a few days ago. 

If I try the following:

```ruby 
db = Sequel.connect("mysql2://user:passwd@host:3306/mydb", :loggers => [Logger.new($stdout)])
columns = db.fetch("SELECT COLUMN_NAME, DATA_TYPE, COLUMN_TYPE FROM INFORMATION_SCHEMA.COLUMNS
 WHERE table_name = 'my_table'").all
ap(columns, indent: -2, index: false)
```
I get following weird output:
```ruby
[
  {
    :COLUMN_NAME => 0,
    :DATA_TYPE   => 0,
    :COLUMN_TYPE => 0
  },
  {
    :COLUMN_NAME => 0,
    :DATA_TYPE   => 0,
    :COLUMN_TYPE => 0
  },
  {
    :COLUMN_NAME => 0,
    :DATA_TYPE   => 0,
    :COLUMN_TYPE => 0
  },
  {
    :COLUMN_NAME => 0,
    :DATA_TYPE   => 0,
    :COLUMN_TYPE => 0
  },
  {
    :COLUMN_NAME => 0,
    :DATA_TYPE   => 0,
    :COLUMN_TYPE => 0
  }
]
```
Maybe someone knows what could be the reason here.

Thanks
-- Armin

**Comment by jeremyevans** [comment]
Try recompiling the mysql2 gem (`gem pristine mysql2`), because it looks like something is very wrong at the driver level.  You could confirm that by using mysql2 directly and see if you get the same results.

**Reply by wuarmin** [reply]
Thank you for the fast response Mr. Evans! `gem pristine mysql2` did not change the behavior.
The error happens here: https://github.com/jeremyevans/sequel/blob/ace201dc4134ac11a3dddc0b630736dc0c78cbd8/lib/sequel/adapters/shared/mysql.rb#L130

It seems, that `fetch('SELECT version()').single_value!` returns the integer `10` instead of the string `10.5.12-MariaDB-0+deb11u1`.

**Comment by jeremyevans** [comment]
For some reason, every data type returned is being treated as integer.  Why that is, I have no idea, but it points to something broken at a level below Sequel.  Can you replicate this issue in a different environment?

**Reply by wuarmin** [reply]
Very strange. The error occurs at my dev-env. At my prod-env (same mysql2-get version) it works. Tomorrow I will check, if I can replicate it in another env.

**Reply by wuarmin** [reply]
I executed this code at dev and prod-env:
```ruby
client = Mysql2::Client.new(:host => "myhost", :username => "me", :password => "abc123", :database => "mydb")
results = client.query("SELECT version()")
results.each do |row|
  ap row
end
```
dev-env-output:
```ruby
{
    "version()" => 10
}
```
prod-env-output:
```ruby
{
    "version()" => "10.5.12-MariaDB-0+deb11u1"
}
```

**Reply by wuarmin** [reply]
I'm on Ubuntu and had to install `libmariadbclient-dev`. Before that only `libmariadb-dev` was installed, as described in the `mysql2`-docs. But then suddenly the error occurred and that was no longer sufficient, so, as I said, I installed the `libmariadbclient-dev`-package.
```txt
apt-get install libmariadbclient-dev
```
After that I recompiled the mysql2 gem and now it works :fireworks: .

**Comment by fnordfish** [comment]
Just to add for the next person running into this.
I ran into the same issue with Rockylinux trying to connect to a MariaDB 11.4.  
Mysql2 will build against `mariadb-devel` or `mariadb-connector-c` but then return only Integer values.

On Rockylinux, I had to use `mysql-devel` instead

**Comment by jeremyevans** [comment]
For users having issues with the `mysql2` adapter due to underlying driver issues, I recommend trying the `trilogy` adapter, as the `trilogy` driver is self-contained and should build the same way regardless of what packages are installed on the system.
